src/transaction/transaction.rs

```diff
@@ -499,16 +503,30 @@ impl Transaction {
     /// # });
     /// ```
     pub async fn commit(&mut self) -> Result<Option<Timestamp>> {
-        if !matches!(
-            self.status,
-            TransactionStatus::StartedCommit | TransactionStatus::Active
-        ) {
-            return Err(Error::OperationAfterCommitError);
+        {
+            let mut status = self.status.write().await;
+            if !matches!(
+                *status,
+                TransactionStatus::StartedCommit | TransactionStatus::Active
+            ) {
+                return Err(Error::OperationAfterCommitError);
+            }
+            *status = TransactionStatus::StartedCommit;
         }
-        self.status = TransactionStatus::StartedCommit;
 
         let primary_key = self.buffer.get_primary_key().await;
         let mutations = self.buffer.to_proto_mutations().await;
+        if mutations.is_empty() {
+            assert!(primary_key.is_none());
+            return Ok(None);
+        }
+
+        if self.options.kind == TransactionKind::Optimistic
+            && self.options.auto_heartbeat
+            && !self.is_heartbeat_started
```

Comment:
The checks can all be moved to `start_auto_heartbeat`. Then maybe rename `start_auto_heartbeat` to reflect the change.

Reply:
What about `may_start_auto_heartbeat`?

Reply:
I suggest `ensure_start_auto_heartbeat`

Reply:
Do you mean `ensure_start_auto_heartbeat` including the `heartbeat` logic or just the check?

Reply:
hmmmm, after deeper thought, I think we may keep it `start_auto_heartbeat`

Reply:
But after the checks have been moved to `start_auto_heartbeat`, `start_auto_heartbeat` may be called multiple times.

Reply:
I think it should be fine.

Reply:
I have moved these checks to `start_auto_heartbeat`, PTAL.